File: Zeze/Transaction/Collections/PSet.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.Immutable;

namespace Zeze.Transaction.Collections
{
    public abstract class PSet<E> : PCollection, ISet<E>
    {
        private readonly Func<ImmutableHashSet<E>, Log> _logFactory;

        protected ImmutableHashSet<E> set;

        protected PSet(long logKey, Func<ImmutableHashSet<E>, Log> logFactory) : base(logKey)
        {
            this._logFactory = logFactory;
            set = ImmutableHashSet<E>.Empty;
        }

        public Log NewLog(ImmutableHashSet<E> value)
        {
            return _logFactory(value);
        }

        public abstract class LogV : Log
        {
            internal ImmutableHashSet<E> Value;

            protected LogV(Bean bean, ImmutableHashSet<E> value) : base(bean)
            {
                Value = value;
            }

            protected void Commit(PSet<E> variable)
            {
                variable.set = Value;
            }
        }

        protected ImmutableHashSet<E> Data
        {
            get
            {
                if (this.IsManaged)
                {
                    var txn = Transaction.Current;
                    if (txn == null)
                    {
                        return set;
                    }
                    txn.VerifyRecordAccessed(this, true);
                    return txn.GetLog(LogKey) is LogV log ? log.Value : set;
                }
                else
                {
                    return set;
                }
            }
        }

        public override string ToString()
        {
            return $"PSet{Data}";
        }

        public int Count => Data.Count;
        public bool IsReadOnly => false;

        public void AddAll(IEnumerable<E> items)
        {
            foreach (var item in items)
                Add(item);
        }

        public abstract bool Add(E item);
        public abstract void Clear();
        public abstract void ExceptWith(IEnumerable<E> other);
        public abstract void IntersectWith(IEnumerable<E> other);
        public abstract bool Remove(E item);
        public abstract void SymmetricExceptWith(IEnumerable<E> other);
        public abstract void UnionWith(IEnumerable<E> other);

        void ICollection<E>.Add(E item)
        {
            this.Add(item);
        }

        public bool Contains(E item)
        {
            return Data.Contains(item);
        }

        public void CopyTo(E[] array, int arrayIndex)
        {
            int index = arrayIndex;
            foreach (var e in Data)
            {
                array[index++] = e;
            }
        }

        public bool IsProperSubsetOf(IEnumerable<E> other)
        {
            return Data.IsProperSubsetOf(other);
        }

        public bool IsProperSupersetOf(IEnumerable<E> other)
        {
            return Data.IsProperSupersetOf(other);
        }

        public bool IsSubsetOf(IEnumerable<E> other)
        {
            return Data.IsSubsetOf(other);
        }

        public bool IsSupersetOf(IEnumerable<E> other)
        {
            return Data.IsSupersetOf(other);
        }

        public bool SetEquals(IEnumerable<E> other)
        {
            return Data.SetEquals(other);
        }

        public bool Overlaps(IEnumerable<E> other)
        {
            return Data.Overlaps(other);
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return Data.GetEnumerator();
        }
        IEnumerator<E> IEnumerable<E>.GetEnumerator()
        {
            return Data.GetEnumerator();
        }

        public ImmutableHashSet<E>.Enumerator GetEnumerator()
        {
            return Data.GetEnumerator();
        }
    }
}

File: ZezeJava/src/Zeze/Transaction/Collections/PSet.java
package Zeze.Transaction.Collections;

import Zeze.Transaction.*;
import java.util.*;
import java.util.stream.Collectors;

import org.pcollections.Empty;

public abstract class PSet<E> extends PCollection {
	private final LogFactory<org.pcollections.PSet<E>> _logFactory;

	protected org.pcollections.PSet<E> set;

	protected PSet(long logKey, LogFactory<org.pcollections.PSet<E>> logFactory) {
		super(logKey);
		this._logFactory = logFactory;
		set = Empty.set();
	}

	public final Log NewLog(org.pcollections.PSet<E> value) {
		return _logFactory.create(value);
	}

	public abstract class LogV extends Log {
		public org.pcollections.PSet<E> Value;

		protected LogV(Bean bean, org.pcollections.PSet<E> value) {
			super(bean);
			Value = value;
		}

		protected final void Commit(PSet<E> variable) {
			variable.set = Value;
		}
	}

	protected final org.pcollections.PSet<E> getData() {
		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			if (txn == null) {
				return set;
			}
			txn.VerifyRecordAccessed(this, true);
			var log = txn.GetLog(LogKey);
			@SuppressWarnings("unchecked")
			var olds = null != log ? ((LogV)log).Value : set;
			return olds;
		}
		return set;
	}

	@Override
	public String toString() {
		return getData().stream().map(Object::toString).collect(Collectors.joining(",", "{", "}"));
	}

	public final int size() {
		return getData().size();
	}

	public final boolean isEmpty() {
		return getData().isEmpty();
	}
	
	public final boolean isReadOnly() {
		return false;
	}

	public abstract boolean add(E item);
	public abstract void addAll(Collection<? extends E> c);
	public abstract void clear();
	public abstract boolean remove(E item);
	public abstract boolean removeAll(Collection<? extends E> c);

	public final boolean contains(Object item) {
		return getData().contains(item);
	}

	public final void CopyTo(E[] array, int arrayIndex) {
		int index = arrayIndex;
		for (var e : getData()) {
			array[index++] = e;
		}
	}
}